dotnet-sdk: Allow multiple packages

baracoder opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Colleagues have switched some projects to version 3.0 of .NET Core. I want my development environment to work with all versions.
This changes allow to combine multiple SDKs and runtimes.

in `configuration.nix`
```
  dotnetCombined = with dotnetCorePackages; combinePackages {
    cli=sdk_3_1_preview;
    withSdks = [ sdk_2_2 sdk_3_0 sdk_3_1_preview ];
    withRuntimes = [ aspnetcore_2_1 sdk_3_0 sdk_2_2 sdk_3_1_preview  ];
  };
```
This setup worked so far. I am open for suggestions.

Using the runtime packages, it should also be possible to build smaller docker images for deployment.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @kuznero 

###### Related issues

https://github.com/NixOS/nixpkgs/issues/73193
https://github.com/NixOS/nixpkgs/pull/72300
https://github.com/NixOS/nixpkgs/pull/69392
https://github.com/NixOS/nixpkgs/pull/69382
https://github.com/NixOS/nixpkgs/pull/69391

baracoder committed:
dotnetCorePackages.sdk_3_0: init at 3.0.100

baracoder committed:
dotnetCorePackages.netcore_2_1: init at 2.1.13

baracoder committed:
dotnetCorePackages.aspnetcore_2_1: init at 2.1.13

baracoder committed:
dotnetCorePackages.sdk_2_1: init at 2.1.509

jonringer commented:
I do like the idea of being able to compose many sdk's. I will have to take a closer look when I get back from work

Elyhaka commented:
I can confirm this is something a Dot Net developer could need : I'm currently working on some legacy projects (2.2) and some new projects that are working with 3.0 only features (Like some Generic Hosts extensions).

I'll will give a try to this PR this week.

reset commented:
I've been using this work in a project of mine with no issues on Ubuntu, Windows WSL2, and a modified version of it for macOS. Excellent work and IMO exactly what I think the dotnet package should look like. @baracoder thank you!

jonringer commented on pkgs/development/compilers/dotnet/default.nix:
```diff
@@ -0,0 +1,46 @@
+    withSdks = [ sdk_2_2 sdk_3_0 sdk_3_1_preview ];
+    withRuntimes = [ aspnetcore_2_1 sdk_3_0 sdk_2_2 sdk_3_1_preview  ];
+  };
+  */
```
may be worthwhile to put this at the top of the file, as that is what people generally see when they first open it

jonringer commented on pkgs/development/compilers/dotnet/combine-packages-builder.sh:
```diff
@@ -0,0 +1,38 @@
+done
+
+
+
```
```suggestion
```

jonringer commented on pkgs/development/compilers/dotnet/combinePackages.nix:
```diff
@@ -0,0 +1,23 @@
+, withRuntimes ? [] 
+, withSdks ? [] 
+}:
+  assert (builtins.length withRuntimes) != 0 || (builtins.length withSdks) != 0;
```
Better user experience:
```suggestion
  lib.assertMsg ((builtins.length withRuntimes) != 0 || (builtins.length withSdks) != 0) 
    ''You must include at least one runtime or sdk, e.g 
      `with dotnetCorePackages; combinePackages {
          withSdks = [ sdk_3_0 ]; 
          withRuntimes = [ aspnetcore_2_1  ];
       };`'';
  ;
```

jonringer reviewed (commented):
on the whole, this looks great, I'll try to compile a few projects, and if it works, then I'll merge it.

Sorry for shelving this so long, was focusing on python for the longest time

FelschR commented:
I've only tried it with v3.1 (stable) of the SDK yet but that has been working without any issues for quite a few project.

Since 3.1 is stable by now it should also be added to the list.
To save you some time this is what I've used:
```
sdk_3_1 = buildNetCoreSdk {
  version = "3.1.100";
  sha512 = "0hvshwsgbm6v5hc1plzdzx8bwsdna2167fnfhxpysqs5mz7crsa4f13m4cxhrbn64lasqz2007nhdrlpgaqvgll6q8736h884aaw5sj";
};
```

baracoder committed:
dotnetCorePakages.sdk_3_1: init at 3.1.100

baracoder commented:
Thank you all for the suggestions. I have updated the branch with the changes and replaced `sdk_3_1_preview` with `sdk_3_1`.
When using Rider, I had to run dotnet clean and invalidate cache to catch up.

evenbrenden commented:
Great! Maybe update the default version in `pkgs/top-level/all-packages.nix` too?
```
-  dotnet-sdk = dotnetCorePackages.sdk_2_2;
+  dotnet-sdk = dotnetCorePackages.sdk_3_1;
```

baracoder commented:
I am not sure about this change. This could break some of the packages, which depend on the current version of dotnet-sdk:
```
% grep -irnl dotnet-sdk
..
pkgs/applications/blockchains/wasabiwallet/default.nix
pkgs/development/python-modules/dotnetcore2/default.nix
pkgs/development/tools/build-managers/msbuild/default.nix
pkgs/servers/nosql/eventstore/default.nix
pkgs/servers/jellyfin/default.nix
pkgs/top-level/python-packages.nix
```

evenbrenden commented:
> I am not sure about this change. This could break some of the packages, which depend on the current version of dotnet-sdk:
> 
> ```
> % grep -irnl dotnet-sdk
> ..
> pkgs/applications/blockchains/wasabiwallet/default.nix
> pkgs/development/python-modules/dotnetcore2/default.nix
> pkgs/development/tools/build-managers/msbuild/default.nix
> pkgs/servers/nosql/eventstore/default.nix
> pkgs/servers/jellyfin/default.nix
> pkgs/top-level/python-packages.nix
> ```

Ah, right. Never mind then - I suppose updating (which includes checking each of these and possibly fixing them to 2.2) warrants a separate issue in any case.

jonringer commented:
@worldofpeace do you see anything out of the ordinary? it LGTM

the reason why such a construction is necessary is:
dotnet makes the assumption that it's able to write to a directory relative to the executable, which ends up being in the nix store, but if the sdk/runtime is already available, then it skips this step. By allowing the sdks/runtimes to be composed through the dotnetPackages.combinePackages function.

jonringer commented:
@baracoder I still haven't forgot this PR, I just keep getting sidetracked with other things :(

jonringer reviewed (commented):
the only thing I noticed is that if i have 2 sdk's, and no runtimes, then this fails to build. It would also be nice to prefer runtimes from sdks if present, as that will reduce closure size

jonringer commented:
the problem with linking to sdk's as "runtimes" is that you don't really get any closure benefit, you bring in the whole other closure once you do that.

jonringer commented on pkgs/development/compilers/dotnet/combine-packages-builder.sh:
```diff
@@ -0,0 +1,36 @@
+done
+
+# Print installed components
+$out/bin/dotnet --info
```
this is what I came up with:
```
source $stdenv/setup

# enumerate the child directories, and link their children
linkSubDirectories() {
    local sourceDir=$1
    local destDir=$2

    for h in `ls $sourceDir`; do
        mkdir -p $destDir/$h
        ln -sf  $sourceDir/$h/* $destDir/$h
    done
}

mkdir -p $out/bin

# CLI will follow real path, so we have to copy it
cp $cli/dotnet $out/
ln -s $out/dotnet $out/bin/

# Create runtime & host links
for runtime in $runtimes; do
    linkSubDirectories $runtime/host $out/host
    linkSubDirectories $runtime/shared $out/shared
done

# Link SDKs
mkdir -p $out/{sdk,host,shared,packs,templates,sdk/NuGetFallbackFolder}
for sdk in $sdks; do
    ln -sf $sdk/templates/* $out/templates/
    ln -sf $sdk/packs/* $out/packs/
    for sd in `ls $sdk/sdk`; do
        [ "$sd" = "NuGetFallbackFolder" ] && continue
        ln -sf $sdk/sdk/$sd $out/sdk/
    done
    linkSubDirectories $sdk/shared $out/shared
    linkSubDirectories $sdk/host $out/host
    # Link NuGetFallbackFolder contents
    test -d $sdk/sdk/NuGetFallbackFolder && find $sdk/sdk/NuGetFallbackFolder -type f -exec ln -sf {} $out/sdk/NuGetFallbackFolder/  \;
done

# Print installed components
$out/bin/dotnet --info
```

jonringer replied:
The main difference is that it will prefer the contents of the sdk's host and shared directories, this gives a few benefits, mainly closure size, but also more runtimes available:

with revisions, an sdk and runtime of the same version won't increase closure size:
```
[08:35:41] jon@jon-desktop /home/jon/projects/nixpkgs ((ea6469760d3...))
$ nix-shell -p "with import ./. {}; (with dotnetCorePackages; combinePackages { withSdks = [ sdk_2_1 ]; withRuntimes = [ netcore_2_1 ]; cli= sdk_2_1; })"
...
.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.13 [/nix/store/bzpxfv0s02ynh11jral0l0l1lsfsv4z2-dotnet-core-combined/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.13 [/nix/store/bzpxfv0s02ynh11jral0l0l1lsfsv4z2-dotnet-core-combined/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.13 [/nix/store/bzpxfv0s02ynh11jral0l0l1lsfsv4z2-dotnet-core-combined/shared/Microsoft.NETCore.App]
...
[nix-shell:/home/jon/projects/nixpkgs]$ nix path-info -Sh /nix/store/bzpxfv0s02ynh11jral0l0l1lsfsv4z2-dotnet-core-combined
/nix/store/bzpxfv0s02ynh11jral0l0l1lsfsv4z2-dotnet-core-combined	 747.2M

$ nix-shell -p "with import ./. {}; (with dotnetCorePackages; combinePackages { withSdks = [ sdk_2_1 ]; withRuntimes = [ ]; cli= sdk_2_1; })"
...
.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.13 [/nix/store/908x63ivkxdfrj4225shsia8jyhyylpj-dotnet-core-combined/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.13 [/nix/store/908x63ivkxdfrj4225shsia8jyhyylpj-dotnet-core-combined/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.13 [/nix/store/908x63ivkxdfrj4225shsia8jyhyylpj-dotnet-core-combined/shared/Microsoft.NETCore.App]
...
[nix-shell:/home/jon/projects/nixpkgs]$ nix path-info -Sh /nix/store/908x63ivkxdfrj4225shsia8jyhyylpj-dotnet-core-combined/
/nix/store/908x63ivkxdfrj4225shsia8jyhyylpj-dotnet-core-combined	 747.2M
```

jonringer replied:
vs the old behavior, it will fail to build without a runtime:

```
$ nix-shell -p "with import ./. {}; (with dotnetCorePackages; combinePackages { withSdks = [ sdk_2_1 ]; withRuntimes = [ netcore_2_1 ]; cli= sdk_2_1; })"
...
.NET Core runtimes installed:
  Microsoft.NETCore.App 2.1.13 [/nix/store/fd2cmglb2vh0rkxbs3h31xm9v83cszfi-dotnet-core-combined/shared/Microsoft.NETCore.App]
...
[nix-shell:/home/jon/projects/nixpkgs]$ nix path-info -Sh /nix/store/fd2cmglb2vh0rkxbs3h31xm9v83cszfi-dotnet-core-combined/
/nix/store/fd2cmglb2vh0rkxbs3h31xm9v83cszfi-dotnet-core-combined	 845.8M

$ nix-shell -p "with import ./. {}; (with dotnetCorePackages; combinePackages { withSdks = [ sdk_2_1 ]; withRuntimes = [ ]; cli= sdk_2_1; })"
these derivations will be built:
  /nix/store/dsdy76rbnd5k62jd56pskx197x0cpgaq-dotnet-core-combined.drv
building '/nix/store/dsdy76rbnd5k62jd56pskx197x0cpgaq-dotnet-core-combined.drv'...
A fatal error occurred, the folder [/nix/store/0pl7pkjx8220yw3h43zjdv5maisrjlp9-dotnet-core-combined/host/fxr] does not exist
builder for '/nix/store/dsdy76rbnd5k62jd56pskx197x0cpgaq-dotnet-core-combined.drv' failed with exit code 131
```

baracoder replied:
As mentioned further down, I have replaced the custom build script with a library function: `buildEnv`, which seams to work well.

baracoder commented:
Thank you for testing and the input.

```
with dotnetCorePackages; combinePakcages { cli=sdk_3_1; withRuntimes=[ sdk_3_1 ]; withSdks=[ sdk_3_1 ]; }
```
Using the same sdk package in runtimes won't increase the closure size, as the directories are just linked.

I had the wrong assumption that explicitly specifying the runtime would decrease closure size for built packages, but this is false, because the cli from the combined package would be used and pull in all SDKs.
`combinePackages` is only be useful for development environments.

I will update the PR tomorrow.

jonringer commented:
also, I don't think the cli argument is necessary, could order the sdks and runtimes first by sdk, then by version, then just choose the first one.

baracoder commented:
I did not have time to update the pull request, but stumbled today on a nice nixpkgs function `buildEnv`. Which makes it look like I have tried to reinvent the wheel:
```
  latestCombined = buildEnv {
    name = "dotnet-combined-latest";
    paths = [ sdk_3_1 sdk_2_2 sdk_3_0 sdk_3_1 ];
    pathsToLink = [ "/host" "/packs" "/sdk" "/shared" "/templates" ];
    ignoreCollisions = true;
    postBuild = ''
      cp ${sdk_3_1}/dotnet $out/dotnet
      mkdir $out/bin
      ln -s $out/dotnet $out/bin/
    '';
  };
```
It works a little bit different but simplifies the process a lot.

worldofpeace reviewed (changes requested):
Created files are missing final newlines.

snewell92 commented:
This change pleases me greatly.

![Little girl with blue cotton candy being so excited at a sporting event.](https://media.giphy.com/media/xTiN0CNHgoRf1Ha7CM/giphy.gif)

baracoder committed:
dotnetPackages: Fix apphost binary

baracoder commented:
Sorry, for the delay, I forgot to ping.

I have updated the merge request. You can now get a combined package like this:
```
with dotnetCorePackages; combinePackages { packages = [ sdk_3_1 sdk_2_2 sdk_3_0 sdk aspnetcore_2_1 ]; }
```
The cli tool will be used from the first entry.

I have smuggled in another commit to fix the `apphost` binary, which is used to create executable files for projects with `OutputType=exe`. There is a catch though: If a library specifies a patch version of the environment like `3.0.1`, it might download a nuget package containing a copy of `apphost`, which is not patched.
```
% find ~/.nuget/packages -iname apphost  
/home/bara/.nuget/packages/microsoft.netcore.app.host.linux-x64/3.0.1/runtimes/linux-x64/native/apphost
```

jonringer commented on pkgs/development/compilers/dotnet/combinePackages.nix:
```diff
@@ -0,0 +1,20 @@
+{ packages }: 
```
since packages is no longer an attr with multiple attributes, how about we just make it a plain array?
```suggestion
packages:
```

jonringer commented on pkgs/development/compilers/dotnet/combinePackages.nix:
```diff
@@ -0,0 +1,20 @@
+    ''You must include at least one package, e.g 
+      `with dotnetCorePackages; combinePackages {
+          packages = [ sdk_3_0 aspnetcore_2_1 ];
+       };`'' ;
```
```

    ''You must include at least one package, e.g 
      `with dotnetCorePackages; combinePackages [ sdk_3_0 aspnetcore_2_1 ];`
    '' ;
```

baracoder replied:
Thank you for the suggestions. I have updated the commits

jonringer commented on pkgs/development/compilers/dotnet/default.nix:
```diff
@@ -0,0 +1,40 @@
+/*
+How to combine packages for use in development:
+dotnetCombined = with dotnetCorePackages; combinePackages { packages = [ sdk_3_1 sdk_2_2 sdk_3_0 sdk aspnetcore_2_1 ]; }
```
```suggestion
dotnetCombined = with dotnetCorePackages; combinePackages [ sdk_3_1 sdk_2_2 sdk_3_0 sdk aspnetcore_2_1 ];
```

jonringer commented on pkgs/development/compilers/dotnet/combinePackages.nix:
```diff
@@ -0,0 +1,20 @@
+{ buildEnv, lib }:
+let cli = builtins.head packages;
+in
+assert lib.assertMsg ((builtins.length packages) != 0) 
```
```suggestion
assert lib.assertMsg ((builtins.length packages) != 0)
```

jonringer commented on pkgs/development/compilers/dotnet/combinePackages.nix:
```diff
@@ -0,0 +1,20 @@
+let cli = builtins.head packages;
+in
+assert lib.assertMsg ((builtins.length packages) != 0) 
+    ''You must include at least one package, e.g 
```
```suggestion
    ''You must include at least one package, e.g
```

jonringer reviewed (commented):
definite improvement, I have a few suggestions

jonringer commented on pkgs/development/compilers/dotnet/combinePackages.nix:
```diff
@@ -0,0 +1,20 @@
+    ''You must include at least one package, e.g 
+      `with dotnetCorePackages; combinePackages {
+          packages = [ sdk_3_0 aspnetcore_2_1 ];
+       };`'' ;
```
You'll probably want to switch the order of the assert and let. Otherwise you'll get the head error from having an empty list rather than the helpful message:

```
assert lib.assertMsg ((builtins.length packages) != 0) 
  ''
    ...
  '';
let
  cli = builtins.head packages;
in
```

baracoder committed:
dotnetCorePackages: Add function to combine SDK packages

Fixup assert


Fixup: Move comment to top


Fixup combine


Fixup combine


Fixup buildDotnet


Fixup default.nix


Fixup combine packages


dotnetCorePackages: Fixup combinePackages

Co-Authored-By: Jon <jonringer@users.noreply.github.com>

jonringer commented:
I'm happy with this for the time being. If it needs to be revisited we can do so in another PR.

This is a much needed improvement to nixpkgs, thanks @baracoder for up-streaming this :)

jonringer commented:
hydra built it https://hydra.nixos.org/eval/1565649#tabs-new . Once a successful release is done, we'll be able to use this on unstable :)

nyanloutre commented:
I made a pull request to add the missing runtimes and use it with Jellyfin : #77961

Jumblemuddle commented:
How would I go about using this `combinePackages` from `nix-env`? Do I need to create a `.nix` file and install that or put it in my nix config instead?

baracoder commented:
@Jumblemuddle I am using `nix repl` to intsall it in my profile:
```
$ nix repl /path/to/nixpkgs
nix-repl> :i with dotnetCorePackages; combinePackages [ sdk_3_1 sdk_2_2 sdk_3_0 aspnetcore_2_1 ]
```
Don't know if it is possible to do `with nix-env`

jonringer commented:
I would probably use a shell.nix in the repository that you need it:

```
with import <nixpkgs> {};

mkShell rec {
  name = "dotnet";
  buildInputs = [
    (with dotnetCorePackages; combinePackages [ sdk_3_1 sdk_2_1 ])
  ];
}
```
doing `nix-shell` will then read in the shell.nix and give you a dotnet with:
```
$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.100
 Commit:    cd82f021f4

Runtime Environment:
 OS Name:     nixos
 OS Version:  20.03pre209250.7184df6beb8
 OS Platform: Linux
 RID:         linux-x64
 Base Path:   /nix/store/cdns908qgng8rk0ngm1dqrd5q1qqnhqa-dotnet-sdk-3.1.100/sdk/3.1.100/

Host (useful for support):
  Version: 3.1.0
  Commit:  65f04fb6db

.NET Core SDKs installed:
  2.1.509 [/nix/store/sdsv8i00h54iii9j8wsrxkjplvmc8z3m-dotnet-core-combined/sdk]
  3.1.100 [/nix/store/sdsv8i00h54iii9j8wsrxkjplvmc8z3m-dotnet-core-combined/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.13 [/nix/store/sdsv8i00h54iii9j8wsrxkjplvmc8z3m-dotnet-core-combined/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.13 [/nix/store/sdsv8i00h54iii9j8wsrxkjplvmc8z3m-dotnet-core-combined/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0 [/nix/store/sdsv8i00h54iii9j8wsrxkjplvmc8z3m-dotnet-core-combined/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.13 [/nix/store/sdsv8i00h54iii9j8wsrxkjplvmc8z3m-dotnet-core-combined/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0 [/nix/store/sdsv8i00h54iii9j8wsrxkjplvmc8z3m-dotnet-core-combined/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

jonringer commented:
once #77961 gets merged, and I write up some documentation, the I'll probably backport this to 19.09 so they can use it as well

baracoder commented:
:man_facepalming: I have made an error in copying things from `src`: `cp ./` will copy `env-vars` which pulls the source `tar.gz` package into the runtime closure
```
% nix why-depends /nix/store/8w7566287q3zalhn0xixl9610ssqirfx-dotnet-sdk-3.1.100 /nix/store/cc94mviqcb0wpqq5488p7a0ggp8v3hny-dotnet-sdk-3.1.100-linux-x64.tar.gz
/nix/store/8w7566287q3zalhn0xixl9610ssqirfx-dotnet-sdk-3.1.100
╚═══env-vars: …=".".declare -x src="/nix/store/cc94mviqcb0wpqq5488p7a0ggp8v3hny-dotnet-sdk-3.1.100-linux-x64.ta…
    => /nix/store/cc94mviqcb0wpqq5488p7a0ggp8v3hny-dotnet-sdk-3.1.100-linux-x64.tar.gz
```
`noDumpEnvVars=true;` should fix this. I will create a PR

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/why-cant-i-install-dotnet-in-my-rust-build-environment-shell/5786/1

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/what-are-your-goals-for-20-03/4773/16
